Use wildcard popup for confirmation

unclejustin opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/32374

This PR replaces the built-in `window.confirm` dialogs with branded modals.

![PixelSnap 2022-04-07 at 20 22 08](https://user-images.githubusercontent.com/1855233/162344729-526d78d9-ef97-4066-b45c-ebda9a8031a4.png)

It does this by adding three new components.

- Shared ConfirmationModal: Generic modal to provide consistent formatting and functionality
- ConfirmDeleteModal: Implementation of ConfirmationModal with delete specific text and handlers
- ConfirmRemoveModal: Implementation of ConfirmationModal with remove specific text and handlers

I noticed that all of the call outs to delete and remove could be co-located with the actual menu items, which cleaned up the implementation quite a bit. Now context menu just triggers these modals which then handle dispatching the requests.

## Test plan

- Navigate to a dashboard with insights
- Use the insight context menu and select "Delete"
- Should see new confirmation modal
- Modal should be cancelable
- Modal should trigger deletion
- Repeat above for remove
## App preview:
- [Link](https://sg-web-32374-code-insights-use-wildcard.onrender.com)

vovakulikov commented on client/shared/src/components/ConfirmationModal.tsx:
```diff
@@ -0,0 +1,30 @@
+import React from 'react'
```
I also see other groups of components that stands outside of this system, something like our context or branch pickers. These components are stateful they have fetching logic and their own state inside, so it something that shouldn'd be inside wildcard component. I would say idea of having shared component has some pros for easy reusing this component in different consumers

For example, in code insights we want to reuse context picker in the dilldown panel. So I would say it makes sense to have let's say `@sourcegraph/pickers` package. cc @sourcegraph/frontend-devs 

```tsx
<ContextPicker 
  contextType={ContextType.query}
  autofocus={true}
  // other ContextPicker specific props
  className={styles.drilldownContextPicker}
/>
```

[Figma Designs](https://www.figma.com/file/Ml5dtV9ni4mfffbfOfF9jg/Allow-search-contexts-as-filters-to-code-insights-(v0)-%2333249?node-id=260%3A11)

<img width="400" alt="Screenshot 2022-04-08 at 12 05 21" src="https://user-images.githubusercontent.com/18492575/162392931-e2b0fe6d-0173-46e5-9965-5f27ea9dba49.png">

unclejustin replied:
I heartily agree with all of the above. I will update accordingly.

vovakulikov commented on client/shared/src/components/ConfirmationModal.tsx:
```diff
@@ -0,0 +1,30 @@
+    const { showModal, handleCancel, handleConfirmation, header, message } = props
+
+    return (
+        <Modal isOpen={showModal} position="center" aria-label="Delete confirmation modal">
```
@unclejustin I think aria-label="Delete confirmation modal" is too specific for this since other consumer might have used this in different scenarios.

unclejustin replied:
Womp! That' was a mistake 😅 fixing.

vovakulikov commented on client/shared/src/components/ConfirmationModal.tsx:
```diff
@@ -0,0 +1,30 @@
+
+export interface ConfirmationModalProps {
+    showModal: boolean
+    handleCancel: () => void
```
@unclejustin I remember that we had a discussion (probably with @limitedmage) somewhere in Slack that it would be cool to have some naming convention around our prop callback naming. The rule that I'm trying to stick is that all public callbacks prop are named with `on<specic callback action>` and all internal handlers inside component are called with `handle<action name>` names.

unclejustin replied:
Solid! Will do.

vovakulikov commented on client/web/src/enterprise/insights/components/insights-view-grid/components/backend-insight/BackendInsight.tsx:
```diff
@@ -84,7 +84,7 @@ export const BackendInsightView: React.FunctionComponent<BackendInsightProps> =
     // Handle insight delete and remove actions
     const { loading: isDeleting, delete: handleDelete } = useDeleteInsight()
-    const { remove: handleRemove, loading: isRemoving } = useRemoveInsightFromDashboard()
+    const { loading: isRemoving } = useRemoveInsightFromDashboard()
```
@unclejustin I think I don't fully understand this hook here. `useRemoveInsightFromDashboard()` hook here exposes loading state variable (inside of `useRemoveInsightFromDashboard` loading is just an internal state that created with useState hook) 

And we do not call here any delete logic (callback) from this hook, therefore this loading state will never be in "true" position. Therefore our UI about deleting/removing insight inside of insight card will never be rendered. 

It's actually a good moment for us to revisit UI about removing and deleting insight. @AlicjaSuska maybe instead of having loading state inside insight card we could have loading state inside "delete insight" confirmation modal and close this modal if an insight was sucessfully deleted and show error inside confirmaton model if it wasn't

unclejustin replied:
Ah dang, you are totally correct! I was expecting that loading would be shared across components. I was wrong. So very wrong #shame.

I'm going to move the "loading" UI into the modal for now and see what Alicja thinks of it.

vovakulikov commented on client/web/src/enterprise/insights/components/insights-view-grid/components/insight-context-menu/ConfirmDeleteModal.tsx:
```diff
@@ -0,0 +1,32 @@
+    showModal,
+    handleCancel,
+}) => {
+    const { delete: handleDelete } = useDeleteInsight()
```
So here we could use loading state from `useDeleteInsight` and for example render loading spinner next to submit button inside confirmation modal and disable closing this model by esc or outside click. @AlicjaSuska I think this might be a good compomise while we don't have a good way to notify user that we've done something (I remember that we had a convo about notification panel for actions like this)

unclejustin replied:
Using loading here now as suggested. Just disabling the buttons and closing for now. Will sync up with Alicja on Monday.

vovakulikov reviewed (commented):
@unclejustin I left a few comments about general things around components and one thing that I would change before merging this PR but it perhaps requires @AlicjaSuska input.

unclejustin commented on client/shared/src/components/ConfirmationModal.tsx:
```diff
@@ -0,0 +1,30 @@
+    showModal: boolean
+    handleCancel: () => void
+    handleConfirmation: () => void
+    header: React.ReactElement | string
```
This is totally fair. I was perhaps being too rigid in my formatting. I agree with moving all of the content into children.

unclejustin commented on client/web/src/enterprise/insights/components/insights-view-grid/components/insight-context-menu/InsightContextMenu.tsx:
```diff
@@ -46,78 +42,94 @@ export const InsightContextMenu: React.FunctionComponent<InsightCardMenuProps> =
+                                    onSelect={() => setShowRemoveConfirm(true)}
+                                    disabled={withinVirtualDashboard}
+                                    data-tooltip={
+                                        withinVirtualDashboard
```
Ooh yeah it should. Guess we missed this one on our first sweep. Done: [#33691](https://github.com/sourcegraph/sourcegraph/issues/33691)

unclejustin commented:
@vovakulikov I've addressed all the issues I can without getting feedback from @AlicjaSuska . Alicja I will try and sync up with you next week to figure out the rest of these issues.

Vova, no further work is needed from you until the design issues are resolved. Thank you for such a thorough review!

unclejustin commented:
@vovakulikov updated with design tweaks from Alicja. Please re-review 🙏 

Note: this still needs a storybook entry before it can be merged in, but I want to wait on that until it passes tech review. Assuming you approve I will add a story for the modals and ping frontend platform team for a final review.

AlicjaSuska commented:
@unclejustin @vovakulikov I left a comment about the copy in the main [issue](https://github.com/sourcegraph/sourcegraph/issues/32374)

unclejustin commented:
@AlicjaSuska your comments have been implemented thank you! 🙇 

@vovakulikov ready for code review

@sourcegraph/frontend-platform I'm proposing adding a component in this PR. Could you kindly review? If you think this should be moved to its own separate PR please let me know ASAP. I'll change this one to be self contained and open up another. Thanks!

vovakulikov commented on client/web/src/enterprise/insights/components/insights-view-grid/components/backend-insight/BackendInsight.tsx:
```diff
@@ -162,8 +158,6 @@ export const BackendInsightView: React.FunctionComponent<BackendInsightProps> =
                             menuButtonClassName="ml-1 d-inline-flex"
                             zeroYAxisMin={zeroYAxisMin}
                             onToggleZeroYAxisMin={() => setZeroYAxisMin(!zeroYAxisMin)}
-                            onRemoveFromDashboard={dashboard => handleRemove({ insight, dashboard })}
```
Nice! I really like how everything is simplified now here.

vovakulikov commented on client/web/src/enterprise/insights/components/insights-view-grid/components/insight-context-menu/InsightContextMenu.tsx:
```diff
@@ -46,78 +42,94 @@ export const InsightContextMenu: React.FunctionComponent<InsightCardMenuProps> =
+                    </>
+                )}
+            </Menu>
+            <ConfirmDeleteModal
```
I think having delete/remove modals close to the triggers that they call makes sense to me. But wondering what it would be if we wanted to generalize modal and confirmation system in our app. Probably idea for wildcard. 

Register all modals in one place and have context based API for calling different modals from any places within the registry modals context. 

```tsx
// app.tsx - the root of code insight app 
const REGISTRY = {
   [CodeInsight.InsightDelete]: ConfirmDeleteModal,
   [CodeInsight.InsightRemove]: ConfirmRemoveModal,
 }
 
 const App = () => (
    <ModalRegestryContext registry= {REGISTRY} >
         // .....
    </ModalRegestryContext>
 )
 
 // ContextMenu.tsx
 
 const ContextMenu = () => {
   const { modal } = useModal()
   // ....
   const handleDeleteInsight = async (values) => {
      await modal.dispatch({ type: CodeInsight.InsightDelete, payload: { insight, values } })
      // If modal confirmation has been canceled this promise will be rejected and we won't fire telemetry service 
      telemetryService.log()
   }
}
```

With this API we have one single place where all consumer specific modals live and we have a clear logic to avoid having local setStates in consumer's components. @sourcegraph/frontend-platform curious what do you think about this?

unclejustin replied:
TLDR: If you feel strongly about it, I'll make it happen.

I'm torn. I don't like moving this up to the main app as it feels a bit "hidden" to me.
BUT this is basically what some of the react toast libraries do.
BUT those are much more simple. You just pass in a message and type.
Here I'm concerned about what that payload ends up being. Think of what the type looks like for dispatch. I mean we could alter the type based on type (hah!) but it seems like a lot of mental hurdles to jump through to display a modal.

So 🤷 

I'm inclined to leave it "inline" with the code because it makes sense to _me_, but I am wide open to others' opinions.

vovakulikov replied:
Just to be clear this thought is just another idea which goes beyond this PR and maybe the thing that we could/should do on app level and this is not code insights specific so feel free to ignore this in this PR

unclejustin replied:
😅 Ah! Got it. Thanks for the clarification!

vovakulikov commented on client/web/src/enterprise/insights/hooks/use-remove-insight.ts:
```diff
@@ -27,19 +29,17 @@ export function useRemoveInsightFromDashboard(): useRemoveInsightFromDashboardAP
             setError(undefined)
 
             try {
+                await sleep(5000)
```
Do we need this? or is this something that sneaked to this PR during testing? @unclejustin

unclejustin replied:
YIKES NO

Thank you for catching this embarrassing debugging code 😅

vovakulikov commented on client/wildcard/src/components/ConfirmationModal/ConfirmationModal.tsx:
```diff
@@ -0,0 +1,43 @@
+    confirmText?: string
+}
+
+export const ConfirmationModal: React.FunctionComponent<React.PropsWithChildren<ConfirmationModalProps>> = props => {
```
I thought `React.FunctionComponent` itself adds children to the props and we don't need to have `React.PropsWithChildren` @unclejustin?

lrhacker replied:
I'd say this is the way to go to prepare for React 18 (we haven't finished upgrading yet, right?), based on the [decision to remove implicit children](https://solverfox.dev/writing/no-implicit-children/)!

unclejustin replied:
Thanks for the commentary @lrhacker! I was unaware of this. All I know is TS was complaining and this was the only thing that would calm it down.

vovakulikov reviewed (approved):
@unclejustin good work, I left a few small comments but besides this it looks great!
